File: app/src/main/java/rogue_opcode/AudioResource.java
// AudioResource.java
// Common audio interface
//
// Copyright © 2010-2017 Brigham Toskin
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <http://code.google.com/p/rogue-op/wiki/LICENSE>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode;

import java.util.HashMap;

import rogue_opcode.soundy.FXR;
import rogue_opcode.soundy.MusicTrack;
import rogue_opcode.soundy.SoundEffect;
import rogue_opcode.soundy.Soundy;


/**
 * Defines a common interface for simple audio objects.
 *
 * @author Brigham Toskin
 */
public abstract class AudioResource
{
	/**
	 * Maps {@code resource} IDs to {@code AudioResource}s. <b>Note: This may
	 * be inefficient with memory for some operations; use carefully!</b>
	 */
	public static HashMap<Integer, AudioResource> sAllARs;
	static
	{
		sAllARs = new HashMap<Integer, AudioResource>();
	}

	protected int mResID;
	protected float mGain;
	protected float mPan;

	// factory functionality ///////////////////////////////////////////////////

	/**
	 * Enumeration of available audio types.
	 */
	public enum AudioType
	{
		/** Indicates a short PCM audio clip suitable for sound effects. */
		EFFECT,
		/** Indicates a longer PCM audio clip suitable for in game music. */
		MUSIC,
		/** Indicates a synthesized audio clip from SFXR. */
		SYNTH_EFFECT,
		/** Indicates a synthesized 4-track song. */
		SYNTH_MUSIC
	}

	/**
	 * Constructs an <code>AudioResource</code> instance, defaulting to the more
	 * expensive but safer implementation to allow for longer audio files.
	 * 
	 * @param pResID audio resource to load.
	 * @return a newly constructed specific instance.
	 * @see AudioResource#ICanHas(int, AudioType)
	 */
	// XXX: does this still make sense?
	public static AudioResource ICanHas(int pResID)
	{
		return ICanHas(pResID, AudioType.MUSIC);
	}

	/**
	 * Constructs an <code>AudioResource</code> instance based on the hint
	 * parameter <code>pType</code>. A limitation of the Android audio
	 * APIs is that the underlying <code>SoundPool</code> objects are ill suited
	 * to playing audio samples that are longer than just a few seconds. As an
	 * alternative, this method can instantiate an implementation that utilizes
	 * <code>MediaPlayer</code> on the back end; this allows for longer audio
	 * clips at the expense of more computational resources.
	 *
	 * @param pResID audio resource to load.
	 * @param pType hint as to the type and characteristics of audio file we
	 *        wish to load; this parameter affects which underlying
	 *        implementation is constructed.
	 * @return a newly constructed specific instance.
	 * @see AudioResource#ICanHas(int)
	 * @see AudioResource.AudioType
	 */
	public static AudioResource ICanHas(int pResID, AudioType pType)
	{
		AudioResource tAudio;
		switch(pType)
		{
		case EFFECT:
			tAudio = new SoundEffect(pResID);
			break;
		case MUSIC:
			tAudio = new MusicTrack(pResID);
			break;
		case SYNTH_EFFECT:
			tAudio = new FXR(pResID);
			break;
		case SYNTH_MUSIC:
			tAudio = new Soundy(pResID);
			break;
		default:
			tAudio = null;
		}
		return tAudio;
	}

	protected AudioResource(int pResID)
	{
		sAllARs.put(pResID, this);
		mResID = pResID;
		mGain = 0.5f;
		mPan = 0.0f;
	}

	abstract protected void die();

	public static void Die()
	{
		// we are shutting down, so we don't care about performance too much
		for(AudioResource tAR : sAllARs.values())
		{
			tAR.die();
		}
	}

	// audio interfaces ////////////////////////////////////////////////////////

	public abstract void Play();
	public abstract void Stop();
	public abstract void Pause();
	public abstract void Resume();

	public float Gain()
	{
		return mGain;
	}
	public void Gain(float pGain)
	{
		if(pGain > 1.0f)
			pGain = 1.0f;
		if(pGain < 0)
			pGain = 0;
		mGain = pGain;
	}
	public void Mute()
	{
		Gain(0);
	}

	public abstract boolean Loop();
	public abstract void Loop(boolean pLoop);

	public float Pan()
	{
		return mPan;
	}
	public void Pan(float pPan)
	{
		if(pPan > 1.0f)
			pPan = 1.0f;
		if(pPan < -1.0f)
			pPan = -1.0f;
		mPan = pPan;

	}

	// static access ///////////////////////////////////////////////////////////

	/**
	 * Finds the {@code AudioResource} associated with the specified {@code
	 * resource} ID, if it exists. You may optionally override this method in
	 * class implementations to create an appropriate instance if the resource
	 * hasn't been previously loaded.
	 *
	 * @param pResID the {@code resource} ID to retrieve.
	 * @return the found {@code AudioResource} instance, or {@code null} if it
	 *         hasn't been loaded.
	 */
	public static AudioResource FindByResID(int pResID)
	{
		return sAllARs.get(pResID);
	}
}

File: lib/src/main/java/rogue_opcode/AudioResource.kt
// AudioResource.kt
// Common audio interface
//
// Copyright © 2010-2017 Brigham Toskin
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <https://github.com/IonoclastBrigham/RogueOp/blob/master/LICENSE.md>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode

import android.support.annotation.AnyRes
import android.util.SparseArray
import rogue_opcode.soundy.FXR
import rogue_opcode.soundy.MusicTrack
import rogue_opcode.soundy.SoundEffect
import rogue_opcode.soundy.Soundy


private operator fun <T> SparseArray<T>.iterator() = object : Iterator<T> {
	private var mIndex = 0

	override fun next() = valueAt(mIndex++)
	override fun hasNext() = mIndex < size()
}

/**
 * *Defines a common interface for simple audio objects.*
 *
 * @author Brigham Toskin
 */
abstract class AudioResource constructor(@AnyRes pResID: Int) {
	companion object {
		/**
		 * Maps `resource` IDs to `AudioResource`s. **Note: This may
		 * be inefficient with memory for some operations; use carefully!**
		 */
		var sAllARs = SparseArray<AudioResource>()

		/**
		 * Constructs an `AudioResource` instance based on the hint
		 * parameter `pType`. A limitation of the Android audio
		 * APIs is that the underlying `SoundPool` objects are ill suited
		 * to playing audio samples that are longer than just a few seconds. As an
		 * alternative, this method can instantiate an implementation that utilizes
		 * `MediaPlayer` on the back end; this allows for longer audio
		 * clips at the expense of more computational resources.

		 * @param pResID audio resource to load.
		 * *
		 * @param pType hint as to the type and characteristics of audio file we
		 * *		wish to load; this parameter affects which underlying
		 * *		implementation is constructed.
		 * *
		 * @return a newly constructed specific instance.
		 * *
		 * @see AudioResource.ICanHas
		 * @see AudioResource.AudioType
		 */
		@JvmOverloads fun ICanHas(pResID: Int, pType: AudioType = AudioType.MUSIC)
			= when(pType) {
				AudioType.EFFECT -> SoundEffect(pResID)
				AudioType.MUSIC -> MusicTrack(pResID)
				AudioType.SYNTH_EFFECT -> FXR(pResID)
				AudioType.SYNTH_MUSIC -> Soundy(pResID)
			}

		fun Die() {
			for(tAR in sAllARs) tAR.die()
		}

		// static access ///////////////////////////////////////////////////////////

		/**
		 * Finds the `AudioResource` associated with the specified `resource` ID, if it exists. You may optionally override this method in
		 * class implementations to create an appropriate instance if the resource
		 * hasn't been previously loaded.
		 * @param pResID the `resource` ID to retrieve.
		 * @return the found `AudioResource` instance, or `null` if it
		 * *		 hasn't been loaded.
		 */
		fun FindByResID(pResID: Int): AudioResource? {
			return sAllARs[pResID]
		}
	}

	/**
	 * Enumeration of available audio types.
	 */
	enum class AudioType {
		/** Indicates a short PCM audio clip suitable for sound effects.  */
		EFFECT,
		/** Indicates a longer PCM audio clip suitable for in game music.  */
		MUSIC,
		/** Indicates a synthesized audio clip from SFXR.  */
		SYNTH_EFFECT,
		/** Indicates a synthesized 4-track song.  */
		SYNTH_MUSIC
	}


	protected var mResID = pResID

	protected var mGain = 0.5f
	protected var mPan = 0f

	init { sAllARs.put(pResID, this) }

	protected abstract fun die()

	// audio interfaces ////////////////////////////////////////////////////////

	abstract fun Play()
	abstract fun Stop()
	abstract fun Pause()
	abstract fun Resume()

	fun Gain() = mGain

	open fun Gain(pGain: Float) {
		mGain = pGain.coerceIn(0f..1.0f)
	}

	fun Mute() = Gain(0f)

	abstract fun Loop(): Boolean
	abstract fun Loop(pLoop: Boolean)

	fun Pan() = mPan

	open fun Pan(pPan: Float) {
		mPan = pPan.coerceIn(-1.0f..1.0f)

	}
}
